src/builtin_path.cpp

```diff
@@ -0,0 +1,770 @@
+    options_t opts;
+    int optind;
+    int retval = parse_opts(&opts, &optind, 0, argc, argv, parser, streams);
+    if (retval != STATUS_CMD_OK) return retval;
+
+    int n_transformed = 0;
+    arg_iterator_t aiter(argv, optind, streams, opts.null_in ? '\0' : '\n');
+    while (const wcstring *arg = aiter.nextstr()) {
+        auto pos = find_extension(*arg);
+
+        if (!pos) {
+            path_out(streams, opts, *arg);
+            continue;
+        }
+
+        // This ends up being empty if the filename ends with ".".
+        // That's arguably correct.
+        //
+        // So we print an empty string but return true,
+        // because there *is* an extension, it just happens to be empty.
+        wcstring ext = arg->substr(0, *pos);
+        if (opts.quiet && !ext.empty()) {
+            // Return 0 if we *had* an extension
+            return STATUS_CMD_OK;
+        }
+        path_out(streams, opts, ext);
+        n_transformed++;
+    }
+
+    return n_transformed > 0 ? STATUS_CMD_OK : STATUS_CMD_ERROR;
+}
+
+static int path_real(parser_t &parser, io_streams_t &streams, int argc, const wchar_t **argv) {
+    options_t opts;
+    int optind;
+    int retval = parse_opts(&opts, &optind, 0, argc, argv, parser, streams);
+    if (retval != STATUS_CMD_OK) return retval;
+
+    int n_transformed = 0;
+    arg_iterator_t aiter(argv, optind, streams, opts.null_in ? '\0' : '\n');
+    while (const wcstring *arg = aiter.nextstr()) {
+        auto real = wrealpath(*arg);
+
+        if (!real) {
+            continue;
+        }
+
+        // Return 0 if we found a realpath.
+        if (opts.quiet) {
+            return STATUS_CMD_OK;
+        }
+        path_out(streams, opts, *real);
+        n_transformed++;
+    }
+
+    return n_transformed > 0 ? STATUS_CMD_OK : STATUS_CMD_ERROR;
+}
+
+static int path_expand(parser_t &parser, io_streams_t &streams, int argc, const wchar_t **argv) {
+    options_t opts;
+    opts.type_valid = true;
+    opts.perm_valid = true;
+    int optind;
+    int retval = parse_opts(&opts, &optind, 0, argc, argv, parser, streams);
+    if (retval != STATUS_CMD_OK) return retval;
+
+    int n_transformed = 0;
+    arg_iterator_t aiter(argv, optind, streams, opts.null_in ? '\0' : '\n');
+    while (const wcstring *arg = aiter.nextstr()) {
+        // TODO: Error handling.
+        // TODO: This is uncancelable.
+        cancel_checker_t cc = [] { return false; };
+        // TODO: Find a good limit.
+        completion_receiver_t out {1024};
+        wcstring pat = parse_util_unescape_wildcards(*arg, true /* force ? */);
+        auto res = wildcard_expand_string(pat, parser.vars().get_pwd_slash(), expand_flag::expanded_wildcards, cc, &out);
+
+        if (opts.quiet && !out.empty()) {
+            return STATUS_CMD_OK;
+        }
+        auto expanded = out.take();
+        std::sort(expanded.begin(), expanded.end(),
```

Comment:
I'm afraid I don't follow?

Reply:
sorry the location was off, my review tool has some rough edges

I mean `for (auto comp : expanded)` should be `for (const auto &comp : expanded)`, otherwise there's an unnecessary copy